Allow to define a custom controller for each entity

javiereguiluz opened the pull request:
Entities now can define an optional `controller` option to set the custom controller executed for them:

```yaml
easy_admin:
    entities:
        # ...
        Product:
            controller: AppBundle\Admin\ProductController
            # ...
```

This will greatly improve the customization of mid-sized backends.

---

The implementation is still WIP and I'll probably change a lot of things. The new doc is also WIP.

javiereguiluz commented:
I need ideas from the community to **help me solve a very strange issue**:

1) After applying this PR, the demo application keeps working as usual.

2) Travis says that tests pass for this PR

![travis_tests](https://cloud.githubusercontent.com/assets/73419/15551889/54d6d938-22b8-11e6-90f3-cf92a96a6d3d.png)

3) However, in my local computer tests are a mess:

![local_tests](https://cloud.githubusercontent.com/assets/73419/15551903/5d720a0e-22b8-11e6-9891-161322c2dc64.png)

---

I'm getting errors like this one:

![error_message](https://cloud.githubusercontent.com/assets/73419/15551911/676c3c50-22b8-11e6-9841-b3b7e1acacef.png)

This is the full text stack trace:

```
[1] InvalidArgumentException: The controller for URI "/admin/" is not callable. Class "JavierEguiluz\Bundle\EasyAdminBundle\Controller\AdminController" does not have a method "__invoke".
    at n/a
        in EasyAdminBundle/vendor/symfony/http-kernel/Controller/ControllerResolver.php line 79

    at Symfony\Component\HttpKernel\Controller\ControllerResolver->getController(object(Request))
        in EasyAdminBundle/vendor/symfony/http-kernel/Controller/TraceableControllerResolver.php line 46

    at Symfony\Component\HttpKernel\Controller\TraceableControllerResolver->getController(object(Request))
        in EasyAdminBundle/vendor/symfony/http-kernel/HttpKernel.php line 127

    at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), '1')
        in EasyAdminBundle/vendor/symfony/http-kernel/HttpKernel.php line 62

    at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), '1', true)
        in EasyAdminBundle/vendor/symfony/http-kernel/Kernel.php line 169

    at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
        in EasyAdminBundle/vendor/symfony/http-kernel/Client.php line 79

    at Symfony\Component\HttpKernel\Client->doRequest(object(Request))
        in EasyAdminBundle/vendor/symfony/framework-bundle/Client.php line 131

    at Symfony\Bundle\FrameworkBundle\Client->doRequest(object(Request))
        in EasyAdminBundle/vendor/symfony/browser-kit/Client.php line 317

    at Symfony\Component\BrowserKit\Client->request('GET', '/admin/?action=list&entity=Category&view=list')
        in EasyAdminBundle/Tests/Fixtures/AbstractTestCase.php line 62

    at JavierEguiluz\Bundle\EasyAdminBundle\Tests\Fixtures\AbstractTestCase->getBackendPage(array('action' => 'list', 'entity' => 'Category', 'view' => 'list'))
        in EasyAdminBundle/Tests/Controller/CustomEntityControllerTest.php line 48

    at JavierEguiluz\Bundle\EasyAdminBundle\Tests\Controller\CustomEntityControllerTest->requestListView()
        in EasyAdminBundle/Tests/Controller/CustomEntityControllerTest.php line 28

    at JavierEguiluz\Bundle\EasyAdminBundle\Tests\Controller\CustomEntityControllerTest->testListAction()
        in  line 

    at ReflectionMethod->invokeArgs(object(CustomEntityControllerTest), array())
        in phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php line 910

    at PHPUnit_Framework_TestCase->runTest()
        in phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php line 770

    at PHPUnit_Framework_TestCase->runBare()
        in phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php line 685

    at PHPUnit_Framework_TestResult->run(object(CustomEntityControllerTest))
        in phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php line 726

    at PHPUnit_Framework_TestCase->run(object(PHPUnit_Framework_TestResult))
        in phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php line 747

    at PHPUnit_Framework_TestSuite->run(object(PHPUnit_Framework_TestResult))
        in phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php line 429

    at PHPUnit_TextUI_TestRunner->doRun(object(PHPUnit_Framework_TestSuite), array('listGroups' => false, 'loader' => null, 'useDefaultConfiguration' => true, 'testSuffixes' => array('Test.php', '.phpt'), 'configuration' => 'EasyAdminBundle/phpunit.xml.dist'))
        in phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php line 155

    at PHPUnit_TextUI_Command->run(array('/usr/local/bin/phpunit', 'EasyAdminBundle/Tests/Controller/CustomEntityControllerTest.php'), true)
        in phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php line 106

    at PHPUnit_TextUI_Command::main()
        in /usr/local/bin/phpunit line 514
```

yceruto commented on Controller/AdminController.php:
```diff
@@ -704,12 +705,21 @@ protected function renderForbiddenActionError($action)
+        } else {
+            $controller = $this->get('easyadmin.controller_resolver')->instantiate($this->entity['controller']);
+            if (method_exists($controller, 'initialize')) {
+                $controller->initialize($this->request);
```
This means that events (`EasyAdminEvents::PRE_INITIALIZE` and `EasyAdminEvents::POST_INITIALIZE`) and `initialize` process will run twice ? instead you can copy the current `AdminController` initialized configuration?

javiereguiluz replied:
Nice catch! In 3cc25bc I've added an `initialized` property which prevents this.

yceruto commented:
I have tried to reproduce this bug locally, but all tests has been passed successfully. :disappointed:

yceruto commented on Controller/ControllerResolver.php:
```diff
@@ -0,0 +1,41 @@
+
+    public function instantiate($controllerClass)
+    {
+        $controller = new $controllerClass();
```
maybe check `AdminController` parent class before that? with a nice exception?

javiereguiluz replied:
I thought about that ... but strictly speaking it's not mandatory to extend from the default AdminController ... but I'm going to add that check to the method which calls `initialize()` if it exists. Thanks for your comment!

yceruto commented on Configuration/ConfigManager.php:
```diff
@@ -186,7 +186,7 @@ private function processConfig()
     {
         $configPasses = array(
-            new NormalizerConfigPass(),
+            new NormalizerConfigPass($this->container->get('easyadmin.controller_resolver')),
```
It has not implemented the `__constructor` for `NormalizerConfigPass`  ?

ogizanagi commented:
@javiereguiluz : Sorry I can't review this fully right now, however one quick question : Why not using a [`kernel.controller`](http://symfony.com/doc/current/components/http_kernel/introduction.html#the-kernel-controller-event) event to change the controller according to the entity config ?

javiereguiluz commented:
@ogizanagi that's a very nice proposal! I'll need to make some tests, specially for these two scenarios that should work:

1) It's not mandatory for the custom controller to extend from the default `AdminController`
2) If it extends from AdminController ... the `initialize()` method should be called automatically without the user having to do so.

javiereguiluz commented:
@yceruto thanks for trying to reproduce the bug locally. I'm still suffering it. Let's see if it persists when I try @ogizanagi's idea.

yceruto commented:
https://github.com/javiereguiluz/EasyAdminBundle/pull/883#issuecomment-190855427 here is an example using `kernel.controller` listener

yceruto commented on Controller/AdminController.php:
```diff
@@ -80,6 +83,10 @@ public function indexAction(Request $request)
      */
     protected function initialize(Request $request)
     {
+        if (true === $this->isInitialized) {
```
I'm not sure that double initialization has been prevented ? there are two controller instances, so `isInitialized` is `false` for each of them ?

javiereguiluz commented:
**Update**: I've changed the code completely according to @yceruto and @ogizanagi comments and using @yceruto's code. Thanks guys! Your solution is infinitely better than mine.

I still suffer the same weird issues with the tests locally ... but Travis says that everything is OK.

I've tweaked the entire documentation and I've added a new "Chapter 7" to explain the override mechanism based on PHP. I think everything is better explained now. Please tell me what you think about it.

Thanks!

javiereguiluz commented:
**Update**: everything is working again on my local computer. The problems were caused by some modifications that I made in the Symfony code inside the `vendor/` directory and which I didn't remember. 😡

yceruto commented:
:+1: LGTM and congradulations for the 1,000 stars! :tada:

javiereguiluz commented:
@yceruto thanks ;) We achieved 1,000 stars yesterday ... but I'm waiting until tomorrow Monday to make some noise on Twitter 😄

vnivuahc commented:
Hi,
I got the following issue when I defined my own entity controller, overriding the default `AdminController` :
> No route found for "GET /admin/"

To avoid this error, I had to override the `indexAction()` method :

```
/**
 * {@inheritdoc}
 */
public function indexAction(Request $request)
{
    return parent::indexAction($request);
}
```
Note that you have to **NOT** add the `@Route` annotation in this method.
(This point made me mad for hours : O)
